eclipse-theia/theia
# Use binary message RPC protocol for plugins

#### What it does

**Note:** Includes the new protocol from https://github.com/eclipse-theia/theia/pull/11011 and, thus, should be merged afterwards.

- Refactor plugin-ext rpc protocol to reuse the new binary message-rpc protocol
  - Remove custom RPC message encoding and handling reuse message-rpc
  - Implement `QueuingChannelMultiplexer` that queues messages and sends them accumulated on the next `process.tick` (replaces the old Multiplexer implementation)
  - Refactors proxy handlers and remote target handlers
  - Use `Channel` instead of `MessageConnection` for creating new instances of `RPCProtocol`
  - Implement special message encoders and decoders for the plugin communication
- Adapt the `HostedPluginServer` and `HostedPluginClient` API to send/receive messages in binary format instead of strings.
- Implement a message protocol for process pipe streams to be able to send messages in binary format between the `hosted-plugin-process` and the `plugin-host`.

Contributed on behalf of STMicroelectronics
Part of https://github.com/eclipse-theia/theia/issues/10684

#### How to test
- Install a VSCode extension and verify that it works as expected, e.g. registered commands can be executed, file changes are applied, etc.

#### Review checklist


#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
